Improve `generators.toPretty`

infinisil opened the pull request:
###### Motivation for this change
For big Nix expressions, `toPretty` can be very hard to read, as it doesn't output any newlines. This PR implements the `multiline` option for `toPretty`, enabled by default, which makes it output newlines with proper indentation. Example output:
```nix
{
  foo = {
    bar = 10;
    baz = [
      "baz"
    ];
  };
}
```

In addition, I made some other improvements:
- Don't quote attribute names if they don't need to be quoted by using `escapeNixIdentifier`, introduced in https://github.com/NixOS/nixpkgs/pull/82461/commits/f9eb3d158a47dccb1e4762f3ed91f224fc96dc7e
- Switch away from the δ and λ symbols for derivations and functions respectively
- Multiline strings are printed correctly now

Ping @Profpatsch @rycee 

Best reviewed commit-by-commit. Specifically there's one commit in there that just refactors the function without changing its behavior.

###### Things done

- [x] Added a new test case for all functionality

infinisil committed:
lib/generators.toPretty: Only quote attribute names if necessary

infinisil committed:
lib/generators.toPretty: Wrap in a go function

As a preparation to the following commit

infinisil committed:
lib/generators.toPretty: Implement multiline printing

Profpatsch commented on lib/generators.nix:
```diff
@@ -203,19 +203,24 @@ rec {
-  }@args: let go = v: with builtins;
+    allowPrettyValues ? false,
+    /* If this option is true, the output is indented with newlines for attribute sets and lists */
+    multiline ? false
```
should we flip to `true` by default? Makes more sense to me. Since it’s intended to be used in debug output, I don’t see how we could break user expectation if we split to multiline by default.

Profpatsch commented on lib/tests/misc.nix:
```diff
@@ -474,6 +474,17 @@ runTests {
+    };
+    expected = rec {
+      list = "[\n  3\n  4\n  [\n    false\n  ]\n]";
+      attrs = "{\n  bar = {\n    foo = \"baz\";\n  };\n  foo = null;\n}";
```
nit: multiline strings here

Profpatsch commented on lib/generators.nix:
```diff
@@ -205,12 +205,14 @@ rec {
-    multiline ? false
+    multiline ? false,
+    /* If multiline is enabled, prefix all new lines with this string */
+    prefix ? ""
```
what do we need this for?

Infinisil replied:
See the PR description for an example of `prefix` usage. Though I'm now realizing that this won't work for multiline strings. I'll see if I can make multiline strings work nicely, probably with `builtins.replaceStrings [ "\n" ] ..` or so

Profpatsch replied:
Hm, I see. It’s probably okay to have this as an option.

Profpatsch reviewed (changes requested):
I love this, I’ve been wanting to have multiline toPretty in forever.

Two additional things:
1) I’d leave the `if/then/else` form instead of going vial `typeOf`, since it’s more straightforward and easier to adjust.
2) I’d render lists and attrsets that only have one element on one line to be slightly more compact. From 2 elements, split onto newlines. Though that might be easier said than done …

infinisil committed:
lib/generators.toPretty: Improved string printing, handling newlines

infinisil committed:
lib/generators.toPretty: Switch away from δ and λ

- These symbols can be confusing for those not familiar with them
- There's no harm in making these more obvious
- Terminals may not print them correctly either

Also changes the function argument printing slightly to be more obvious

infinisil committed:
lib/generators.toPretty: Print [] and {} compactly

infinisil committed:
lib/generators.toPretty: functors should print as functions

Not attribute sets. So move the function case forward

Infinisil commented:
Did some changes:
- Multiline strings are printed correctly now
- Switch away from the δ and λ symbols for derivations and functions respectively
- Went back to the big `if else`. I don't think it's nicer, but it works out well for another change:
- To print `__functor`'s as functions
- Removed the `prefix` argument again. This can be achieved with `builtins.replaceStrings [ "\n" ] [ "\nprefix" ]` as well

I decided to not render single-element lists/attrs without newlines, because I fear this makes it more complicated to understand the output, as the indentation can't be used to see the nesting level anymore.

Here's a sample output for this Nix expression:
```nix
{
  value = {
  fun = lib.id;
  funargs = lib.types.submoduleWith;
  multiline = {
    multi = ''
      hello
      there!
    '';
    multi' = ''
      hello
      there!'';
    single = "hello there!";
  };
  pkg = pkgs.hello;
  single = { foo = [ { bar = [ null ]; } ]; };
}
```

Which renders as

```
{
  fun = <function>;
  funargs = <function, args: {modules, shorthandOnlyDefinesConfig?, specialArgs?}>;
  multiline = {
    multi = ''
      hello
      there!
    '';
    multi' = ''
      hello
      there!'';
    single = "hello there!";
  };
  pkg = <derivation /nix/store/2nnf3klixf4j2ih8x0zx4albfzlk74l9-hello-2.10.drv>;
  single = {
    foo = [
      {
        bar = [
          null
        ];
      }
    ];
  };
}
```

Profpatsch reviewed (approved):
Looks very good to me :+1: 

Haven’t reviewed the logic in-depth, but we have tests so it’s fine. LGTM

edolstra commented:
I suppose eventually we'd want this to be a builtin function in Nix, since it's needed in the UI anyway, and it's easier on the C++ side to handle circular values etc.
